domain/secret/service/access.go

```diff
@@ -186,29 +191,42 @@ func (s *SecretService) RevokeSecretAccess(ctx context.Context, uri *secrets.URI
 		p.SubjectTypeID = domainsecret.SubjectModel
 	}
 
-	return s.secretState.RevokeAccess(ctx, uri, p)
+	return withCaveat(ctx, func(innerCtx context.Context) error {
+		return s.secretState.RevokeAccess(innerCtx, uri, p)
+	})
 }
 
-// canManage checks that the accessor can manage the secret.
-// If the request is for a secret owned by an application, the unit must be the leader.
-func (s *SecretService) canManage(
+// getManagementCaveat returns a function within which an operation can be
+// executed if the caveat remains satisfied.
+// If the secret is unit-owned and the unit can manage it, the caveat is always
+// permissive.
+// If the secret is application-owned, the unit must be, and remain the leader
+// of that application.
```

Comment:
question: how do we guarantee the unit remains the leader?

Reply:
This does not make that guarantee yet, it only makes the one check prior. If you check the logic for `WithLease` (I'll rename it to `WithLeadership` tomorrow). you can see how that cancels a context if the lease is lost. I'll thread that in in the patch to follow, which will give us the guarantee.

Reply:
That's done with the following code https://github.com/juju/juju/blob/07e5a0f93fb642555a1e3c35f7a6c9a4249cf89e/domain/leaseservice.go#L30